veps/sig-storage/vm-object-graph.md

```diff
@@ -0,0 +1,279 @@
+# VEP #41: Object Graph API for VM Dependencies
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] : https://github.com/kubevirt/enhancements/issues/41
+
+## **Overview**
+
+This is a proposal to include an `Object Graph` API in KubeVirt to represent VM and VMI dependencies and their relationships.
+
+## **Motivation**
+
+As new features continue to be added to KubeVirt, the graph of objects related to VMs (DataVolumes, PersistentVolumeClaims, InstanceTypes, Preferences, Secrets, ConfigMaps, etc.) continues to expand. Identifying all the objects that a VM depends on for tasks like backup, disaster recovery, or migration can be error-prone. We should simplify this process for users and partners by creating an authoritative way to retrieve a structured object graph.
+
+## **Goals**
+
+- Introduce a new Object Graph API to represent the list of resources a VM or VMI depends on.
+- Expose this API as a subresource of VirtualMachines and VirtualMachineInstances.
+- Provide a flexible and extensible data structure.
+- Allow for basic filtering (present or future).
+
+## **Non-Goals**
+
+- Reimplementation of existing VM/VMI specs.
+- Building a generic Kubernetes-wide graph system.
+
+## **User Stories**
+
+1. As a KubeVirt user, I want a clear way to retrieve all VM and VMI-related dependencies.
+2. As a backup partner, I want a way to identify a list of a VM's related objects so I can comprehensively backup and restore everything a VM needs.
+3. As a VM owner, I want to easily define an ACM-discovered application and protect my VM with disaster recovery software.
+4. As a VM owner, I want to migrate my VM from one cluster to another and identify all necessary dependencies for replication.
+5. As a KubeVirt developer, I want a specific place to keep the object graph code updated when I introduce code that changes the relationship of a VM to its dependent objects.
+
+## **Repos**
+
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [KubeVirt Velero Plugin](https://github.com/kubevirt/kubevirt-velero-plugin)
+
+## **Design**
+
+### API Endpoints
+
+```
+/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachines/{name}/objectgraph
+/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachineinstances/{name}/objectgraph
+```
+
+### API Examples
+
+#### **1. First option: Hierarchical Representation**
```

Comment:
Can you please add a section with your analysis and your preference for option 1 vs option 2? I don't think there are many parent-child relationship right? I mean, it is almost everything under volumes. But in general I tend to prefer a hierarchical representations.